pkg/virt-controller/watch/vmi.go

```diff
@@ -861,6 +867,50 @@ func (c *VMIController) hotplugPodsReady(vmi *virtv1.VirtualMachineInstance, vir
 	return true, nil
 }
 
+func hasMaintenancePodPVC(guestfsPods []*k8sv1.Pod, pvc string) bool {
+	for _, pod := range guestfsPods {
+		if pod.ObjectMeta.Labels[guestfs.VirtGuestfsPVCLabel] == pvc {
+			return true
+		}
+	}
+	return false
+}
+
+// checkMaintenancePodActiveOnVolumes checks if one of the volumes of the VMI is in used by a pod created by virtctl guestfs
+func (c *VMIController) checkMaintenancePodActiveOnVolumes(vmi *virtv1.VirtualMachineInstance) error {
+	objs, err := c.podInformer.GetIndexer().ByIndex(cache.NamespaceIndex, vmi.Namespace)
+	if err != nil {
+		return err
+	}
+	var guestfsPods []*k8sv1.Pod
+	for _, obj := range objs {
+		pod := obj.(*k8sv1.Pod)
+		if pod.ObjectMeta.Labels[virtv1.AppLabel] == guestfs.VirtGuestfsName {
+			guestfsPods = append(guestfsPods, pod)
```

Comment:
There's an optimization here. You can create a map of pvcs being accessed by guestfs pods and avoid the n^2 operation.

so something like

```
pvcsInUse := make(map[string]interface{})
.
.
.
     if pod.ObjectMeta.Labels[virtv1.AppLabel] == guestfs.VirtGuestfsName {
        pvcsInUse[pod.ObjectMeta.Labels[guestfs.VirtGuestfsPVCLabel]] = pod
     }
.
.
.

for _, v := range vmi.Spec.Volumes {
.
.
.
    _, exists := pvcsInUse[claimName]
    if exists {
        return fmt.Errorf("volume %s is in use by virtctl guestfs", claimName)
    }
}

```

Reply:
You're right, thanks! Done